perf: towards a faster `conda run` via inline environment activation

agriyakhetarpal opened the pull request:
<!-- Hello! Thanks for submitting a PR! To help make things go a bit more
     smoothly, we would appreciate it if you follow this template. -->

### Description

<!-- Good things to put here include:
       - reasons for the change (please link any relevant issues!),
       - any noteworthy (or hacky) choices to be aware of,
       - or what the problem resolved here looked like. -->

This PR is a part of #14993, and is an effort to make `conda run` faster. Our approach here is to avoid calling `conda activate` in the wrapper script and instead use platform-specific activators. On POSIX platforms, we build the `PosixActivator` and have it write the activation actions to the wrapper script. For Windows, we disable file streaming for `CmdExeActivator` (save I/O overhead) and have it print the activation output to memory, whose lines we then iterate over and `CALL`/`SET` as needed. On Unix, it's even possible for us to run the commands as `sh -c "<...>"` instead of using a temporary file (which should be a little faster), but that's a platform-specific optimisation, and I've not proceeded with it for now.

I was optimistic with my expectations for results here, and they are pretty promising, too. I've used [`hyperfine`](https://github.com/sharkdp/hyperfine) for benchmarking, as follows:

<details>
<summary><h3>Benchmarking</h3></summary>

#### Running a command in the base environment; in this case, `python --version`

##### Before

```bash
hyperfine --warmup 10 --runs 1000 "conda run -- python --version"
Benchmark 1: conda run -- python --version
  Time (mean ± σ):      1.262 s ±  0.179 s    [User: 0.919 s, System: 0.303 s]
  Range (min … max):    1.118 s …  3.736 s    1000 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
```

##### After

```bash
hyperfine --warmup 10 --runs 1000 "conda run -- python --version" 
Benchmark 1: conda run -- python --version
  Time (mean ± σ):      1.022 s ±  0.334 s    [User: 0.716 s, System: 0.245 s]
  Range (min … max):    0.880 s …  7.534 s    1000 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 ```

which is a 19.02% speedup! 🔥

#### Test suite

Even running parts of the test suite shows nice results. Here is the output:

##### Before

```bash
hyperfine "pytest -k test_main_run.py" --warmup 5 --runs 100
Benchmark 1: pytest -k test_main_run.py
  Time (mean ± σ):     25.888 s ±  2.872 s    [User: 17.240 s, System: 5.871 s]
  Range (min … max):   22.594 s … 38.291 s    100 runs
 ```

##### After

```bash
hyperfine "pytest -k test_main_run.py" --warmup 5 --runs 100 
Benchmark 1: pytest -k test_main_run.py
  Time (mean ± σ):     19.559 s ±  3.858 s    [User: 12.411 s, System: 4.326 s]
  Range (min … max):   16.203 s … 36.211 s    100 runs
```

Let's hope that CodSpeed also provides us with similar numbers 😉

</details>

### Checklist - did you ...

<!-- If any of the following items aren't relevant to your contribution,
     please either tick them or use ~strikethrough~ so we know you've gone
     through the checklist. -->

- [x] Add a file to the `news` directory ([using the template](https://github.com/conda/conda/blob/main/news/TEMPLATE)) for the next release's release notes?
     <!-- All "significant" changes should get an entry:
            - user-facing changes or enhancements
            - bug fixes
            - deprecations
            - documentation updates
            - etc -->
- [ ] ~Add / update necessary tests?~ Not needed; there is no functional/behaviour change
- [ ] ~Add / update outdated documentation?~


<!-- Just as a reminder, everyone in all conda org spaces (including PRs)
     must follow the Conda Org Code of Conduct (link below).

     Finally, once again, thanks for your time and effort. If you have any
     feedback in regards to your experience contributing here, please
     let us know!

     Helpful links:
       - Conda Org COC: https://github.com/conda/conda/blob/main/CODE_OF_CONDUCT.md
       - Contributing docs: https://github.com/conda/conda/blob/main/CONTRIBUTING.md -->

agriyakhetarpal commented:
Argh; I should have run the full test suite locally instead of a few files – looks like there is some work ahead for me. I'll mark this as a draft for a bit until I fix the failures.

soapy1 commented on conda/utils.py:
```diff
@@ -245,7 +240,54 @@ def wrap_subprocess_call(
+            conda_meta_dir.mkdir(parents=True, exist_ok=True)
+
+            try:
+                _context.changeps1 = False
```
What is the motivation for setting this `changeps1` configuration for this section of code? 
I don't think it's really good to be modifying the context like this. Another option might be to use this [fresh_context](https://github.com/conda/conda/blob/main/conda/base/context.py#L2077).

jaimergp replied:
cc @agriyakhetarpal: I share the same concerns as Sophia. Why is this needed at all?

agriyakhetarpal replied:
This was required because I was trying to fix a failing test. I did that prior to the new year, so let me look up what the failure was...

agriyakhetarpal replied:
(I marked my above two comments as off-topic; I'm conflating issues.)

I recall that I did this to perform a micro-optimisation indeed: we don't need to bother with shell prompt modifications, as activation happens in a non-interactive subprocess script and changing the prompt has no visible effect. It does not result in a measurable improvement. I'll drop this.

Side note: I poked around with `fresh_context` a bit after this comment, and it turns out it's quite heavyweight because it reinitialises the entire context.

jaimergp commented on conda/utils.py:
```diff
@@ -245,7 +240,54 @@ def wrap_subprocess_call(
+            # ensure that a sentinel conda-meta directory exists to pass
+            # activation validation.
+            conda_meta_dir = Path(prefix) / "conda-meta"
+            conda_meta_dir.mkdir(parents=True, exist_ok=True)
```
I agree with Sophia here. The behavior in `main` is clearly a bug.

jaimergp commented on conda/activate.py:
```diff
@@ -335,8 +334,6 @@ def _build_activate_stack(self, env_name_or_prefix, stack):
 
                 raise EnvironmentLocationNotFound(prefix)
-        elif env_name_or_prefix in (RESERVED_ENV_NAMES):
-            prefix = context.root_prefix
```
Nice opt. This is being done in `locate_prefix_by_name` already 🚀

jaimergp commented on conda/utils.py:
```diff
@@ -245,7 +240,54 @@ def wrap_subprocess_call(
+                activate_env = activator.activate()
+            finally:
+                _context.__dict__.pop("changeps1", None)
+                _context.__dict__.pop("dev", None)
```
Hm, why this? This is a destructive action for the global context.

jaimergp commented on conda/utils.py:
```diff
@@ -245,7 +240,54 @@ def wrap_subprocess_call(
+                key, value = line.split("=", 1)
+                fh.write(f'{silencer}SET "{key}="\n') if value == "" else fh.write(
+                    f'{silencer}SET "{key}={value}"\n'
+                )
```
Isn't this the same? If `value == ""`, then the `SET` line is still `"{key}="`.

```suggestion
                fh.write(f'{silencer}SET "{key}={value}"\n')
```

jaimergp commented on conda/utils.py:
```diff
@@ -245,7 +240,54 @@ def wrap_subprocess_call(
+
+                activator = activator_cls(activator_args)
+                activator._parse_and_set_args()
+                # CmdExeActivator writes an .env file by default; let's force in-memory output here
```
I think there were security reasons the .env file was needed. Otherwise the script got blocked by some AV software or something like that. @kenodegard may be able to confirm.

agriyakhetarpal commented on conda/utils.py:
```diff
@@ -314,7 +356,32 @@ def wrap_subprocess_call(
+                _context.__dict__.pop("changeps1", None)
+                _context.__dict__.pop("dev", None)
+
+            fh.write(activate_code)
```
Good point – it did not; but now does via e48d591ee6ffa370fcc3aab67d5d06d835c7f038. Thanks!

soapy1 commented on news/15534-inline-env-activation-conda-run:
```diff
@@ -0,0 +1,19 @@
+
+### Bug fixes
+
+* <news item>
```
Probably also worth it to include a note about the bug fix here too.

```suggestion
* Ensure an appropriate error code is returned when `conda run` is invoked with an invalid environment name or prefix. (#15534)
```

soapy1 reviewed (approved):
Looks great to me! Nice speed up 🤸

agriyakhetarpal commented on tests/core/test_initialize.py:
```diff
@@ -482,6 +482,7 @@ def test_install_conda_sh(verbose):
 def test_install_conda_fish(verbose):
     with tempdir() as conda_temp_prefix:
         target_path = join(conda_temp_prefix, "etc", "fish", "conf.d", "conda.fish")
+        context.dev = False
```
The `test_install_conda_sh` test does it, but the corresponding tests for other shells were not doing so – this also fixes the tests that failed here: https://github.com/conda/conda/actions/runs/20972959121/job/60282210013?pr=15534

agriyakhetarpal replied:
Yes, indeed, considering that this is test code that has been there for long, I think it should be fine... I can switch to using the monkeypatch + reset_context if that helps, sure! I'll also see if there are other places we can clean this up. :)

agriyakhetarpal commented:
I think the problem with these tests is that the parameter values are in `_cache_`. When `reset_context()` is called, the cache might not be properly cleared. I spent some time on this today, but I'm still trying to wrap my head around it... If I don't find it, I'd suggest we go back to setting `context.dev` because that's how we've had it in those tests for a long time...

kenodegard commented on conda/utils.py:
```diff
@@ -254,7 +248,41 @@ def wrap_subprocess_call(
+            activate_env = activator.activate()
+
+            for line in activate_env.splitlines():
+                line = line.rstrip("\r")
```
`str.splitlines` is supposed to already cover this:

https://docs.python.org/3/library/stdtypes.html#str.splitlines

kenodegard commented on news/15534-inline-env-activation-conda-run:
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* `conda run` activates environments inline instead of using a subprocess, thus speeding up the runtime (#15534)
```
```suggestion
* Speedup `conda run` by inlining environment activation instead of using a subprocess. (#15534)
```

agriyakhetarpal commented on conda/utils.py:
```diff
@@ -254,7 +248,41 @@ def wrap_subprocess_call(
+                    fh.write(f'{silencer}CALL "{script}"\n')
+                    # If any of these calls to the activation hook scripts fail, we want
+                    # to exit the wrapper immediately and abort `conda run` right away.
+                    fh.write(f"{silencer}IF %ERRORLEVEL% NEQ 0 EXIT /b %ERRORLEVEL%\n")
```
Yes, I think we should, though it would be a breaking change – even if it was always broken in that sense. I can change it.

agriyakhetarpal replied:
Added in 9323177666e6088e3a6ac41045f063fdf3b67bdc

kenodegard replied:
I went ahead and pulled this into a new PR #15616 since it was related to the recent #15602 and is somewhat separate from this PR

agriyakhetarpal commented:
> > I also want us to get rid of the `--dev` option, is that something we can stop supporting for `conda run`?
> 
> Yes, I think so, as I've had a few run-ins with it since it has broken some tests over the course of this PR. Is it okay to leave that for a subsequent PR?

Looks like I found a TODO for this 😃: https://github.com/conda/conda/blob/97b8bfdf4caa3c75a934cdcbaa39200e7c21d8c0/conda/testing/fixtures.py#L273-L274

beckermr commented on tests/core/test_path_actions.py:
```diff
@@ -189,6 +189,10 @@ def test_CompileMultiPycAction_noarch_python(prefix: Path):
     create_link(sys.executable, python_full_path, LinkType.softlink)
 
+    # create conda-meta directory like _execute_actions does in real installation
+    conda_meta_dir = join(prefix, "conda-meta")
+    mkdir_p(conda_meta_dir)
+
```
This addition to the test suite is suspicious as well. One thing I found when deugging locally is this error in the messages:

```
DirectoryNotACondaEnvironmentError: The target directory exists, but it is not a conda environment.
Use 'conda create' to convert the directory to a conda environment.
  target directory: /home/runner/miniconda3/envs/test
```

That happens when the conda-meta dir is missing. From the change here, it appears that with the changfes in this PR suddenly this had to be done by hand? I wonder if we coded around a bug here instead of fixing it.

cc @jaimergp @agriyakhetarpal

beckermr commented:
OK @jaimergp @agriyakhetarpal Per my tests here, this PR is breaking linking+activation in a way that is not captured by the test suite. My guess is that the test suite envs are too simple. You can see my test of this here https://github.com/LSSTDESC/descwl-shear-sims/pull/238 where I patched this PR out live in the CI on a failing case and now it works.

beckermr commented:
@conda/conda-maintainers I strongly suggest we revert this PR and push a bugfix release ASAP. The bug here is a fundamental one in post-install linking and environment activation. 

These are core functions that have now been broken for ~3 months in a case that is common enough it was found twice independently. Further people are now patching their CI around avoiding conda versions 26.* which is the kind of thing folks used to do many years ago when conda was less stable.

agriyakhetarpal commented:
Hi @beckermr, thanks for raising this. I'll get on my computer and look at figuring out what the problem is on an immediate basis.

agriyakhetarpal commented:
From my preliminary investigations, it looks like this is related to the `set -e` changes in the activation hook calls, i.e., the "fail fast if activation fails" logic. I am moving away from my keyboard, but I will be back in a few hours to implement a fix and see if it fully resolves the linked issues. Edit: I think I am making progress and on the right path.

Also, I'll add a test that creates a conda environment with a package that has an `activate.d` script containing a benignly-failing command, plus a package with a post-link script.

jaimergp commented:
A hypothesis worth considering is that having added `set -e` is revealing an error in activation that had been failing silently for a bit, so the actual root cause may be somewhere else. Food for thought.

beckermr commented:
@jaimergp I don't think there has been a fundamental error, but I could be wrong of course.

Instead, what I think is happening is that bash/sh doesn't have good support for `try...except` idioms. So when the fail fast was added via `set -e`, things are no longer being cleaned up properly. This would explain the `set -e` might be at issue and the fact that in my failing test case, I saw an activation script being run with the wrong prefix.

beckermr commented:
@jaimergp @agriyakhetarpal one thing to remember is that this utility function is used in multiple different contexts in the code base, not just for conda run.

beckermr commented:
I think my hunch is looking close to correct.

Calling conda activate has a line (https://github.com/conda/conda/blob/main/conda/shell/etc/profile.d/conda.sh#L30) that does

```
ask_conda="$(PS1="${PS1:-}" __conda_exe shell.posix "$@")" || \return
```

~and this happens in a subshell so the calling process is not effected.~

~This is effectively a `try...except` idiom in bash, and that idiom was removed when the activate functionality was inlined.~

Edit: I don't think this is quite right actually, but something related it happening.

agriyakhetarpal commented:
I am trying to verify whether my hypotheses are correct with `gdk-pixbuf`, but I am seeing increased retries and occasional failures when accessing https://conda.anaconda.org/conda-forge/osx-arm64/repodata.json, so I can't test reliably. (I don't think it has to do with my internet connection).

The reproducer `conda create -n myenv -c hep-forge -c conda-forge root root-plus root-guard` as mentioned in https://github.com/conda-forge/gdk-pixbuf-feedstock/issues/61 seems to be `linux-64` specific and doesn't seem to work on `osx-arm64`. Would `conda create -n test-gdk-pixbuf -c conda-forge gdk-pixbuf librsvg -y` be an okay reproducer for `osx-arm64`, or should I be attempting a heavier solve?

beckermr commented:
I have not been able to reproduce anywhere but a very heavy CI run here: https://github.com/LSSTDESC/descwl-shear-sims/pull/238.

So I am just using that to test things by patching conda directly. :/

beckermr commented:
@jaimergp @agriyakhetarpal Removing the set commands fixes the issue. 

So I think the fix is in three parts. 

1. Back out this PR to fix the issue. A 20% gain in performance is not worth the bug. 
   Further, while removing the set commands fixes things, we
    - do not understand well enough what the expected behavior was originally to make changes here in the first place
    - do not have a good test case for this bug
    - do not fully understand the bug since it seems to also require big environments (and thus possibly involve a race condition?)

   Thus I do not think a further PR on top of this one is prudent.
2. Recognize that this utility function is used across the code base for different operations. In particular, it is used for both running linking scripts and for `conda run`. These operations may need different error handling semantics.
3. Once we understand 2, put in a proper optimization.

beckermr commented:
So again @conda/conda-maintainers let me very strongly state that we should revert this PR and push a bugfix release ASAP. The bug here is a fundamental one in post-install linking and environment activation, and further our collective understanding of the code base is what caused this bug to pass code review and be shipped in the first place. We should recognize that and be much more conservative in putting in changes into these operations.

agriyakhetarpal commented:
Thanks, @beckermr. I am still facing installation errors, so perhaps it's something to do with the CDN and my region, India. I've written up a patch file below with something. Could you please apply it and try it out there?

```patch
From 0ffbcd230c9c73ea482fac3cee499699e6f28824 Mon Sep 17 00:00:00 2001
From: Agriya Khetarpal <74401230+agriyakhetarpal@users.noreply.github.com>
Date: Fri, 10 Apr 2026 16:46:29 +0530
Subject: [PATCH] Testing something

---
 conda/activate.py |  4 +---
 conda/utils.py    | 29 ++++++++++++++++++++++-------
 2 files changed, 23 insertions(+), 10 deletions(-)

diff --git a/conda/activate.py b/conda/activate.py
index 8556c5e9a..675703018 100644
--- a/conda/activate.py
+++ b/conda/activate.py
@@ -1035,9 +1035,7 @@ class CmdExeRunActivator(CmdExeActivator):
     export_var_tmpl = 'SET "%s=%s"'
     path_var_tmpl = export_var_tmpl
     set_var_tmpl = export_var_tmpl
-    # If any of these calls to the activation hook scripts fail, we want
-    # to exit the wrapper immediately and abort `conda run` right away.
-    run_script_tmpl = 'CALL "%s"\nIF %%ERRORLEVEL%% NEQ 0 EXIT /b %%ERRORLEVEL%%'
+    run_script_tmpl = 'CALL "%s"'
 
 
 class FishActivator(_Activator):
diff --git a/conda/utils.py b/conda/utils.py
index 39f6a26c1..bbf98a45b 100644
--- a/conda/utils.py
+++ b/conda/utils.py
@@ -238,17 +238,26 @@ def wrap_subprocess_call(
             old_path = environ.get("PATH", "")
             if condabin_dir not in old_path.split(os.pathsep):
                 environ["PATH"] = f"{condabin_dir}{os.pathsep}{old_path}"
+
+            # Clear activation state so the activator sees a clean
+            # environment (CONDA_SHLVL=0). Otherwise it inherits the
+            # conda process's own activation state and (my hunch: may) compute
+            # wrong PATH entries or emit spurious deactivate scripts.
+            saved_env = {}
+            for key in ("CONDA_SHLVL", "CONDA_PREFIX", "CONDA_DEFAULT_ENV"):
+                if key in environ:
+                    saved_env[key] = environ.pop(key)
+
             try:
                 activator = activator_cls(activator_args)
                 activator._parse_and_set_args()
                 activate_script = activator.activate()
             finally:
                 environ["PATH"] = old_path
+                environ.update(saved_env)
 
             for line in activate_script.splitlines():
                 fh.write(f"{silencer}{line}\n")
-
-            fh.write(f"{silencer}IF %ERRORLEVEL% NEQ 0 EXIT /b %ERRORLEVEL%\n")
             if debug_wrapper_scripts:
                 fh.write("echo *** environment after *** 1>&2\n")
                 fh.write("SET 1>&2\n")
@@ -294,9 +303,6 @@ def wrap_subprocess_call(
             raise Exception("No compatible shell found!")
 
         with Utf8NamedTemporaryFile(mode="w", delete=False) as fh:
-            # If any of these calls to the activation hook scripts fail, we want
-            # to exit the wrapper immediately and abort `conda run` right away.
-            fh.write("set -e\n")
             if dev_mode:
                 from . import CONDA_SOURCE_ROOT
 
@@ -323,19 +329,28 @@ def wrap_subprocess_call(
             old_path = environ.get("PATH", "")
             if condabin_dir not in old_path.split(os.pathsep):
                 environ["PATH"] = f"{condabin_dir}{os.pathsep}{old_path}"
+
+            # Clear activation state so the activator sees a clean
+            # environment (CONDA_SHLVL=0). Otherwise it inherits the
+            # conda process's own activation state and (my hunch: may) compute
+            # wrong PATH entries or emit spurious deactivate scripts.
+            saved_env = {}
+            for key in ("CONDA_SHLVL", "CONDA_PREFIX", "CONDA_DEFAULT_ENV"):
+                if key in environ:
+                    saved_env[key] = environ.pop(key)
+
             try:
                 activator = activator_cls(activator_args)
                 activator._parse_and_set_args()
                 activate_code = activator.activate()
             finally:
                 environ["PATH"] = old_path
+                environ.update(saved_env)
 
             fh.write(activate_code)
 
             if debug_wrapper_scripts:
                 fh.write(">&2 echo '*** environment after ***'\n>&2 env\n")
-            # Disable exit-on-error for the user's command so we can capture its exit code.
-            fh.write("set +e\n")
             if multiline:
                 # The ' '.join() is pointless since mutliline is only True when there's 1 arg
                 # still, if that were to change this would prevent breakage.
-- 
2.53.0

```

agriyakhetarpal commented:
Ah, looks like you posted some further comments, and GitHub's UI didn't refresh. Please allow me some time to read through them and respond. Thanks!

beckermr commented:
I am happy to test the patch in the mean time while you all digest my comments. :) 

I need to run for an hour or so, but will reply here with the results of my test. I need to ensure I have conda 26.3.1 in my base env in the test as well for this patch above. Right now my patch is removing the set commands from conda 26.1.* and so did not test changes to your subsequent PR.

agriyakhetarpal commented:
Thank you for testing it!

I read your comments, and I agree that it is important to unblock your and @meiyasan's work. I'll create a PR to the 26.3.x patch line that reverts both [#15534](https://github.com/conda/conda/pull/15534) (this PR) and #15672 (my follow-up to this PR).

Next, I'll put together another PR that combines those two PRs, and applies the fixes from today, provided that they end up working correctly for you, based on your test. That PR can go through stricter peer review, until we are comfortable with merging it. I'd request you to look at that PR as well, at your convenience!

beckermr commented:
@agriyakhetarpal One thing to note is that when `conda shell.posix` is called, a new python process is created with `context.__init__()` called again. I suspect this bit is related to the subsequent fixes for global state changes that were put in once the hook function was removed. 

See the code here: 

https://github.com/conda/conda/blob/7dde4c75e13b57efcc369055ba9be7771bc53a5f/conda/cli/main.py#L77

This points also relates to the encapsulation issues I mentioned in that other PR. To keep things inlined and the same, you have to put in all of the changes in the current python process to make it look like it is a new process with `context.__init__()` called. Doing that in a way that is not fragile may require some level of work in the `context` object to allow it to deinit or reinit. Someone will need to go look.

beckermr commented:
> Thanks, @beckermr. I am still facing installation errors, so perhaps it's something to do with the CDN and my region, India. I've written up a patch file below with something. Could you please apply it and try it out there?
> 
> ```diff
> From 0ffbcd230c9c73ea482fac3cee499699e6f28824 Mon Sep 17 00:00:00 2001
> From: Agriya Khetarpal <74401230+agriyakhetarpal@users.noreply.github.com>
> Date: Fri, 10 Apr 2026 16:46:29 +0530
> Subject: [PATCH] Testing something
> 
> ---
>  conda/activate.py |  4 +---
>  conda/utils.py    | 29 ++++++++++++++++++++++-------
>  2 files changed, 23 insertions(+), 10 deletions(-)
> 
> diff --git a/conda/activate.py b/conda/activate.py
> index 8556c5e9a..675703018 100644
> --- a/conda/activate.py
> +++ b/conda/activate.py
> @@ -1035,9 +1035,7 @@ class CmdExeRunActivator(CmdExeActivator):
>      export_var_tmpl = 'SET "%s=%s"'
>      path_var_tmpl = export_var_tmpl
>      set_var_tmpl = export_var_tmpl
> -    # If any of these calls to the activation hook scripts fail, we want
> -    # to exit the wrapper immediately and abort `conda run` right away.
> -    run_script_tmpl = 'CALL "%s"\nIF %%ERRORLEVEL%% NEQ 0 EXIT /b %%ERRORLEVEL%%'
> +    run_script_tmpl = 'CALL "%s"'
>  
>  
>  class FishActivator(_Activator):
> diff --git a/conda/utils.py b/conda/utils.py
> index 39f6a26c1..bbf98a45b 100644
> --- a/conda/utils.py
> +++ b/conda/utils.py
> @@ -238,17 +238,26 @@ def wrap_subprocess_call(
>              old_path = environ.get("PATH", "")
>              if condabin_dir not in old_path.split(os.pathsep):
>                  environ["PATH"] = f"{condabin_dir}{os.pathsep}{old_path}"
> +
> +            # Clear activation state so the activator sees a clean
> +            # environment (CONDA_SHLVL=0). Otherwise it inherits the
> +            # conda process's own activation state and (my hunch: may) compute
> +            # wrong PATH entries or emit spurious deactivate scripts.
> +            saved_env = {}
> +            for key in ("CONDA_SHLVL", "CONDA_PREFIX", "CONDA_DEFAULT_ENV"):
> +                if key in environ:
> +                    saved_env[key] = environ.pop(key)
> +
>              try:
>                  activator = activator_cls(activator_args)
>                  activator._parse_and_set_args()
>                  activate_script = activator.activate()
>              finally:
>                  environ["PATH"] = old_path
> +                environ.update(saved_env)
>  
>              for line in activate_script.splitlines():
>                  fh.write(f"{silencer}{line}\n")
> -
> -            fh.write(f"{silencer}IF %ERRORLEVEL% NEQ 0 EXIT /b %ERRORLEVEL%\n")
>              if debug_wrapper_scripts:
>                  fh.write("echo *** environment after *** 1>&2\n")
>                  fh.write("SET 1>&2\n")
> @@ -294,9 +303,6 @@ def wrap_subprocess_call(
>              raise Exception("No compatible shell found!")
>  
>          with Utf8NamedTemporaryFile(mode="w", delete=False) as fh:
> -            # If any of these calls to the activation hook scripts fail, we want
> -            # to exit the wrapper immediately and abort `conda run` right away.
> -            fh.write("set -e\n")
>              if dev_mode:
>                  from . import CONDA_SOURCE_ROOT
>  
> @@ -323,19 +329,28 @@ def wrap_subprocess_call(
>              old_path = environ.get("PATH", "")
>              if condabin_dir not in old_path.split(os.pathsep):
>                  environ["PATH"] = f"{condabin_dir}{os.pathsep}{old_path}"
> +
> +            # Clear activation state so the activator sees a clean
> +            # environment (CONDA_SHLVL=0). Otherwise it inherits the
> +            # conda process's own activation state and (my hunch: may) compute
> +            # wrong PATH entries or emit spurious deactivate scripts.
> +            saved_env = {}
> +            for key in ("CONDA_SHLVL", "CONDA_PREFIX", "CONDA_DEFAULT_ENV"):
> +                if key in environ:
> +                    saved_env[key] = environ.pop(key)
> +
>              try:
>                  activator = activator_cls(activator_args)
>                  activator._parse_and_set_args()
>                  activate_code = activator.activate()
>              finally:
>                  environ["PATH"] = old_path
> +                environ.update(saved_env)
>  
>              fh.write(activate_code)
>  
>              if debug_wrapper_scripts:
>                  fh.write(">&2 echo '*** environment after ***'\n>&2 env\n")
> -            # Disable exit-on-error for the user's command so we can capture its exit code.
> -            fh.write("set +e\n")
>              if multiline:
>                  # The ' '.join() is pointless since mutliline is only True when there's 1 arg
>                  # still, if that were to change this would prevent breakage.
> -- 
> 2.53.0
> ```

This patch works per CI run: https://github.com/LSSTDESC/descwl-shear-sims/actions/runs/24245679484/job/70791233919?pr=238

So we are getting closer. However, see my recent comments on encapsulation of the various global/activation state changes. We need to do this in a more structured way.
